# A Comparison Analysis of CAO points for 2019, 2020, 2021
Source: [CAO Webpage](http://www.cao.ie/index.php?page=mediastats)
***

## Table of Contents

#### 1. Introduction

#### 2. Retrieving the Data

#### 3. Concatenating & Joining the Data Sets

<br>

### Import Libraries

***



In [600]:
# Import libraries
import pandas as pd

# Create plots.
import matplotlib.pyplot as plt

# Nice plot style. 
import seaborn as sns

# Numerical operations. 
import numpy as np

# Regular expressions.
import re

# HTTP requests.
import requests as rq

# To get dates and times.
import datetime as dt

# Opening URLs.
import urllib.request as urlrq

# Engine to read in excel file.
import openpyxl as oxl

# Ensures plots are shown.
%matplotlib inline

<br>

### Funtion to retrieve current date and time
***

In [601]:
def time():
    # Gets the current date and time
    cur_time = dt.datetime.now()
    # Format as a string
    current_time = cur_time.strftime('%Y%m%d_%H%M%S')
    return current_time

<br>

# 2021 CAO data

http://www.cao.ie/index.php?page=points&p=2021&bb=points

***


<br>

When first starting this report, the 2021 points were only available in a HTML format in the form of round 1 and round 2 data. Later an excel file became available with data specifying the Interview/Portfolio and EOS midpoints. Because of this, the first part of this section will demonstrate the extraction of the round 1 and round 2 data from the HTML, while the latter part of this section will demonstrate the retrieval of Interview/Portfolio and EOS Midpoints columns from the excel file. 

<br>

### Retrieve data from webserver
***

In [602]:
# Retrieves CAO points from the webserver.
response = rq.get('http://www2.cao.ie/points/l8.php')

# Response 200 signifies a successful request/response.
response

<Response [200]>

<br>

### Save the original data
***

In [603]:
# Creates a file path for the original data
pathHTML = 'data/cao2021' + time() + '.html'# Note the importance of the filename and how it
                                              # will be easy to find in folders/sorted appropriately

In [604]:
# Saves the original html file.
with open(pathHTML, 'w') as f:
    f.write(response.text)

<br>

### Charset error on server

***


Technically, server states decoding as:

```
    Content-Type: text/html; charset=iso-8859-1.
``` 

However, one line uses \x96 which isn't defined in iso-8859-1. 

Therefore, we use the similar decoding standard cp1252, which is very similar but includes \x96. 

In [605]:
# Server uses incorrect encoding
orig_encoding = response.encoding

# Corrected encoding to cp1252
response.encoding = 'cp1252'

<br>

### Using regular expressions to extract desired data
***

To do: Explain what the regular expression is doing. COnsider doing step-by-step explanation. 

In [606]:
# Compiles the regular expression for matching lines so it doesn't recompile repeatedly.
re_course = re.compile(r'([A-Z]{2}[0-9]{3})  (.*)')  # 'r' python treats string as raw string and doesnt evaluate back slashes
                                                                    # \ {character} means we want the literal character ie., *
                                                                    # ? means 0 or 1 of 
                                                                    # + means 1 or more of 

<br>

### Iterating through the response line by line
***

To do:

Here we are dealing with the hash and astrisks denoting portfolio and random selection consecutively.

make sure to spell out what hash and * mean. 

Include image as example. 

In [607]:
# Function to separate the *, # and digits in the points string. 

def points_to_list(string):
    portfolio = ''
    if string[0] == '#':
        portfolio = '#'
    random = ''
    if string[-1] == "*":
        random = '*'
    points = ''
    for i in string:
        # Extracts only the digits from the string. 
        if i.isdigit():
            points += i
    return [points, portfolio, random]

In [608]:
# Create a path for the CSV file.
path2021 = 'data/cao2021' + time() + '.csv'

# Keeping count of the courses we are processing.
course_count = 0

# Open the CSV file for writing.
with open(path2021, 'w') as f:
    # Loops through and prints data from response line by line.
    for line in response.iter_lines():
        # Decoding turns bytes into code points and those code points can be displayed on the screen
        d_line = line.decode('cp1252')
        # Match the string specified in re_course, returning only the courses from the response
        if re_course.fullmatch(d_line):
            # Adds one to the course count
            course_count += 1
            # Extract course code and strip of any white space.
            course_code = d_line[:5].strip()
            # Extract course title and strip it of any white space.
            course_title = d_line[7:57].strip()
            # Points.
            course_points = re.split(' +' , d_line[60:])
            # The last course created 3 substrings w/ line split above & the last substring was subsequently removed.
            # Removing substrings that are not useful. 
            if len(course_points) != 2:
                course_points = course_points[:2]
            # Rejoin the substrings with commas. Because course points is a list we need to specify both items. 
            line_split = [course_code, course_title, course_points[0], course_points[1]]
            f.write(','.join(line_split) + '\n')

            # https://web.microsoftstream.com/video/89e78275-d944-444e-bed7-d964a3bd2c35 30.00


<br>

After writing the data to file, I went into the CSV and created a row with headings for each column. 

In [609]:
# Read in 2021 csv file and add header row.
df_2021 = pd.read_csv("data/cao202120211115_190905.csv", 
                names=["course_code", "course_title", "rnd_1", "rnd_2"])    

df_2021['rnd_1'] = df_2021['rnd_1'].to_numpy()

df_2021

course_code                                       course_title rnd_1 rnd_2
0         AL801  Software Design for Virtual Reality and Gaming...   300   NaN
1         AL802  Software Design in Artificial Intelligence for...   313   NaN
2         AL803  Software Design for Mobile Apps and Connected ...   350   NaN
3         AL805  Computer Engineering for Network Infrastructur...   321   NaN
4         AL810  Quantity Surveying                            ...   328   NaN
..          ...                                                ...   ...   ...
944       WD211  Creative Computing                            ...   270   NaN
945       WD212  Recreation and Sport Management               ...   262   NaN
946       WD230  Mechanical and Manufacturing Engineering      ...   230   230
947       WD231  Early Childhood Care and Education            ...   266   NaN
948       WD232  Business Information Systems                  ...   261   261

[949 rows x 4 columns]

<br>

## Get EOS points & portfolio/interview data
***

[read_excel() documentation](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.read_excel.html)

In [610]:
# Save file path to a variable.
eos_excel = 'data/CAOPointsCharts2021.xlsx'

In [611]:
# Read in data.
eos_2021 = pd.read_excel(eos_excel, skiprows=11, engine='openpyxl') # Use openpyxl to open xlsx spreadsheet.

# Check data.
eos_2021.head()

CATEGORY (ISCED Description)                     Course Title  \
0  Engineering and engineering trades  Music and Instrument Technology   
1                              Health              Pharmacy Technician   
2                              Health                   Dental Nursing   
3     Biological and related sciences                  Applied Science   
4         Business and administration                         Business   

  Course Code R1 Points R1 Random R2 Points  R2 Random EOS Points EOS Random  \
0       AL605       211       NaN        NaN       NaN        211        NaN   
1       AL630       308       NaN        NaN       NaN        308        NaN   
2       AL631       311       NaN        NaN       NaN        311        NaN   
3       AL632       297       NaN        NaN       NaN        297        NaN   
4       AL650       AQA       NaN        AQA       NaN        AQA        NaN   

  EOS Midpoints  Course Level                              HEI Test/Interview  \
0           319             6  Athlone Institute of Technology            NaN   
1           409             6  Athlone Institute of Technology            NaN   
2           400             6  Athlone Institute of Technology            NaN   
3           454             6  Athlone Institute of Technology            NaN   
4           351             6  Athlone Institute of Technology            NaN   

   AVP    v  
0  NaN  NaN  
1  NaN  NaN  
2  NaN  NaN  
3  NaN  NaN  
4  avp  NaN

<br>

### Drop columns
***

In [612]:
# Delete irrelevant columns.
eos_2021 = eos_2021.drop(columns=['Course Title', 'CATEGORY (ISCED Description)', 'R1 Points', 'R1 Random', 'R2 Points ', 'R2 Random', 'EOS Points', 'EOS Random', 'Course Level', 'HEI', 'AVP', 'v'])

eos_2021


Course Code EOS Midpoints Test/Interview
0          AL605           319            NaN
1          AL630           409            NaN
2          AL631           400            NaN
3          AL632           454            NaN
4          AL650           351            NaN
...          ...           ...            ...
1446       WD211           392            NaN
1447       WD212           304            NaN
1448       WD230           361            NaN
1449       WD231           366            NaN
1450       WD232           329            NaN

[1451 rows x 3 columns]

<br>

### Concatenate and join tables
***

In [613]:
# Set the index to be the course code as this is what we will join the dataframes on.
eos_2021.set_index('Course Code', inplace=True)
df_2021.set_index('course_code', inplace=True)

In [614]:
# Join dataframes.
df_2021 = df_2021.join(eos_2021)

In [615]:
# Check the dataframe.
df_2021

course_title rnd_1 rnd_2  \
course_code                                                                  
AL801        Software Design for Virtual Reality and Gaming...   300   NaN   
AL802        Software Design in Artificial Intelligence for...   313   NaN   
AL803        Software Design for Mobile Apps and Connected ...   350   NaN   
AL805        Computer Engineering for Network Infrastructur...   321   NaN   
AL810        Quantity Surveying                            ...   328   NaN   
...                                                        ...   ...   ...   
WD211        Creative Computing                            ...   270   NaN   
WD212        Recreation and Sport Management               ...   262   NaN   
WD230        Mechanical and Manufacturing Engineering      ...   230   230   
WD231        Early Childhood Care and Education            ...   266   NaN   
WD232        Business Information Systems                  ...   261   261   

            EOS Midpoints Test/Interview  
course_code                               
AL801                 359            NaN  
AL802                 381            NaN  
AL803                 398            NaN  
AL805                 381            NaN  
AL810                 377            NaN  
...                   ...            ...  
WD211                 392            NaN  
WD212                 304            NaN  
WD230                 361            NaN  
WD231                 366            NaN  
WD232                 329            NaN  

[949 rows x 5 columns]

In [616]:
# Count the rows in the dataframe.
len(df_2021)

949

On the 12-11-2021, it was verified that 949 courses were documented in the 2021 CAO data online at the following webpage http://www2.cao.ie/points/l8.php. This corresponds with the number of courses in our dataset ensuring that there was no loss of data.
***

<br>

# 2020 CAO data

[2020 data from CAO website](http://www.cao.ie/index.php?page=points&p=2020)
***

In [617]:
# Save the URL in a variable.
url_2020 = 'http://www2.cao.ie/points/CAOPointsCharts2020.xlsx'

<br>

### Save the original file
***

In [618]:
# Create a file path for the original data set.
pathxlsx = 'data/CAO2020' + time() + '.xlsx'

# Opening URL
urlrq.urlretrieve(url_2020, pathxlsx)

('data/CAO202020211208_193305.xlsx', <http.client.HTTPMessage at 0x126d7b580>)

<br>

### Load data with pandas
[Pandas documentation for reading in excel data](https://pandas.pydata.org/docs/reference/api/pandas.read_excel.html)
***

The function from the pandas library ot read in an excel file is `pandas.read_excel()`. 

One of this function's parameters is `engine`. If we set the value to openpyxl, we can use this library to open newer excel file formats.

In [619]:
# Load and parse spread sheet. 
df_2020 = pd.read_excel(pathxlsx, skiprows=10, engine='openpyxl') # Use openpyxl to open xlsx spreadsheet.
df_2020

CATEGORY (i.e.ISCED description)  \
0                           Business and administration   
1                         Humanities (except languages)   
2                                                  Arts   
3                                                  Arts   
4                                                  Arts   
...                                                 ...   
1459                       Manufacturing and processing   
1460  Information and Communication Technologies (ICTs)   
1461  Information and Communication Technologies (ICTs)   
1462                                  Personal services   
1463                 Engineering and engineering trades   

                                           COURSE TITLE COURSE CODE2  \
0                                International Business        AC120   
1                                          Liberal Arts        AC137   
2      First Year Art & Design (Common Entry,portfolio)        AD101   
3     Graphic Design and Moving Image Design (portfo...        AD102   
4     Textile & Surface Design and Jewellery & Objec...        AD103   
...                                                 ...          ...   
1459                          Manufacturing Engineering        WD208   
1460                       Software Systems Development        WD210   
1461                                 Creative Computing        WD211   
1462                    Recreation and Sport Management        WD212   
1463           Mechanical and Manufacturing Engineering        WD230   

     R1 POINTS R1 Random * R2 POINTS R2 Random*       EOS EOS Random *  \
0          209         NaN       NaN        NaN       209          NaN   
1          252         NaN       NaN        NaN       252          NaN   
2     #+matric         NaN       NaN        NaN  #+matric          NaN   
3     #+matric         NaN       NaN        NaN  #+matric          NaN   
4     #+matric         NaN       NaN        NaN  #+matric          NaN   
...        ...         ...       ...        ...       ...          ...   
1459       188         NaN       NaN        NaN       188          NaN   
1460       279         NaN       NaN        NaN       279          NaN   
1461       271         NaN       NaN        NaN       271          NaN   
1462       270         NaN       NaN        NaN       270          NaN   
1463       253         NaN       NaN        NaN       253          NaN   

     EOS Mid-point  ...  avp    v Column1 Column2 Column3  Column4  Column5  \
0              280  ...  NaN  NaN     NaN     NaN     NaN      NaN      NaN   
1              270  ...  NaN  NaN     NaN     NaN     NaN      NaN      NaN   
2         #+matric  ...  NaN  NaN     NaN     NaN     NaN      NaN      NaN   
3         #+matric  ...  NaN  NaN     NaN     NaN     NaN      NaN      NaN   
4         #+matric  ...  NaN  NaN     NaN     NaN     NaN      NaN      NaN   
...            ...  ...  ...  ...     ...     ...     ...      ...      ...   
1459           339  ...  NaN  NaN     NaN     NaN     NaN      NaN      NaN   
1460           337  ...  NaN  NaN     NaN     NaN     NaN      NaN      NaN   
1461           318  ...  NaN  NaN     NaN     NaN     NaN      NaN      NaN   
1462           349  ...  NaN  NaN     NaN     NaN     NaN      NaN      NaN   
1463           369  ...  NaN  NaN     NaN     NaN     NaN      NaN      NaN   

      Column6  Column7  Column8  
0         NaN      NaN      NaN  
1         NaN      NaN      NaN  
2         NaN      NaN      NaN  
3         NaN      NaN      NaN  
4         NaN      NaN      NaN  
...       ...      ...      ...  
1459      NaN      NaN      NaN  
1460      NaN      NaN      NaN  
1461      NaN      NaN      NaN  
1462      NaN      NaN      NaN  
1463      NaN      NaN      NaN  

[1464 rows x 23 columns]

In [620]:
# Delete irrelevant columns.
df_2020.drop(columns=['COURSE TITLE', 
                    'CATEGORY (i.e.ISCED description)', 
                    'R1 Random *', 
                    'R2 Random*',
                    'LEVEL',
                    'EOS', 
                    'EOS Random *',  
                    'HEI', 
                    'avp', 
                    'v', 
                    'Column1', 'Column2', 'Column3', 'Column4', 'Column5', 'Column6', 
                    'Column7', 'Column8'])


COURSE CODE2 R1 POINTS R2 POINTS EOS Mid-point Test/Interview #
0           AC120       209       NaN           280              NaN
1           AC137       252       NaN           270              NaN
2           AD101  #+matric       NaN      #+matric                #
3           AD102  #+matric       NaN      #+matric                #
4           AD103  #+matric       NaN      #+matric                #
...           ...       ...       ...           ...              ...
1459        WD208       188       NaN           339              NaN
1460        WD210       279       NaN           337              NaN
1461        WD211       271       NaN           318              NaN
1462        WD212       270       NaN           349              NaN
1463        WD230       253       NaN           369              NaN

[1464 rows x 5 columns]

In [621]:
# Checking random row to ensure data integrity.
df_2020.iloc[650]

CATEGORY (i.e.ISCED description)                                             Arts
COURSE TITLE                        Arts (Drama, Theatre and Performance Studies)
COURSE CODE2                                                                GY118
R1 POINTS                                                                     451
R1 Random *                                                                   NaN
R2 POINTS                                                                     NaN
R2 Random*                                                                    NaN
EOS                                                                           451
EOS Random *                                                                  NaN
EOS Mid-point                                                                 492
LEVEL                                                                           8
HEI                                        National University of Ireland, Galway
Test/Interview #

In [622]:
# Checking the last row.
df_2020.iloc[-1]


CATEGORY (i.e.ISCED description)          Engineering and engineering trades
COURSE TITLE                        Mechanical and Manufacturing Engineering
COURSE CODE2                                                           WD230
R1 POINTS                                                                253
R1 Random *                                                              NaN
R2 POINTS                                                                NaN
R2 Random*                                                               NaN
EOS                                                                      253
EOS Random *                                                             NaN
EOS Mid-point                                                            369
LEVEL                                                                      8
HEI                                        Waterford Institute of Technology
Test/Interview #                                                         NaN

In [623]:
# Saving the data to a CSV file.
path2020 = 'data/cao2020_' + time() + '.csv'
#df_2020.to_csv(path2020, names=["area", "course_title", "course_code", "Mid-point"])

df_2020

CATEGORY (i.e.ISCED description)  \
0                           Business and administration   
1                         Humanities (except languages)   
2                                                  Arts   
3                                                  Arts   
4                                                  Arts   
...                                                 ...   
1459                       Manufacturing and processing   
1460  Information and Communication Technologies (ICTs)   
1461  Information and Communication Technologies (ICTs)   
1462                                  Personal services   
1463                 Engineering and engineering trades   

                                           COURSE TITLE COURSE CODE2  \
0                                International Business        AC120   
1                                          Liberal Arts        AC137   
2      First Year Art & Design (Common Entry,portfolio)        AD101   
3     Graphic Design and Moving Image Design (portfo...        AD102   
4     Textile & Surface Design and Jewellery & Objec...        AD103   
...                                                 ...          ...   
1459                          Manufacturing Engineering        WD208   
1460                       Software Systems Development        WD210   
1461                                 Creative Computing        WD211   
1462                    Recreation and Sport Management        WD212   
1463           Mechanical and Manufacturing Engineering        WD230   

     R1 POINTS R1 Random * R2 POINTS R2 Random*       EOS EOS Random *  \
0          209         NaN       NaN        NaN       209          NaN   
1          252         NaN       NaN        NaN       252          NaN   
2     #+matric         NaN       NaN        NaN  #+matric          NaN   
3     #+matric         NaN       NaN        NaN  #+matric          NaN   
4     #+matric         NaN       NaN        NaN  #+matric          NaN   
...        ...         ...       ...        ...       ...          ...   
1459       188         NaN       NaN        NaN       188          NaN   
1460       279         NaN       NaN        NaN       279          NaN   
1461       271         NaN       NaN        NaN       271          NaN   
1462       270         NaN       NaN        NaN       270          NaN   
1463       253         NaN       NaN        NaN       253          NaN   

     EOS Mid-point  ...  avp    v Column1 Column2 Column3  Column4  Column5  \
0              280  ...  NaN  NaN     NaN     NaN     NaN      NaN      NaN   
1              270  ...  NaN  NaN     NaN     NaN     NaN      NaN      NaN   
2         #+matric  ...  NaN  NaN     NaN     NaN     NaN      NaN      NaN   
3         #+matric  ...  NaN  NaN     NaN     NaN     NaN      NaN      NaN   
4         #+matric  ...  NaN  NaN     NaN     NaN     NaN      NaN      NaN   
...            ...  ...  ...  ...     ...     ...     ...      ...      ...   
1459           339  ...  NaN  NaN     NaN     NaN     NaN      NaN      NaN   
1460           337  ...  NaN  NaN     NaN     NaN     NaN      NaN      NaN   
1461           318  ...  NaN  NaN     NaN     NaN     NaN      NaN      NaN   
1462           349  ...  NaN  NaN     NaN     NaN     NaN      NaN      NaN   
1463           369  ...  NaN  NaN     NaN     NaN     NaN      NaN      NaN   

      Column6  Column7  Column8  
0         NaN      NaN      NaN  
1         NaN      NaN      NaN  
2         NaN      NaN      NaN  
3         NaN      NaN      NaN  
4         NaN      NaN      NaN  
...       ...      ...      ...  
1459      NaN      NaN      NaN  
1460      NaN      NaN      NaN  
1461      NaN      NaN      NaN  
1462      NaN      NaN      NaN  
1463      NaN      NaN      NaN  

[1464 rows x 23 columns]

On 8 December 2021, it was verified that there were 1464 courses listed on the CAO webpage for the year 2020. This corresponds with the number of courses in our data set. 

<br>

# 2019 CAO data
 
http://www2.cao.ie/points/lvl8_19.pdf

***


## Convert pdf to csv
***

What did I do to prepare data before reading in csv file below?

Copy and pasted pdf in Preview, pasted into a Word document so it formats nicely.

Then copied the data from the Word document to a csv file, while deleting preamble and unnecessary data such as preamble, page numbers and full Higher Education INstitution's names while keeping the course code and points etc. 

In [624]:
# Read in 2019 cao csv file & use the tab character as the delimiter.
df_2019 = pd.read_csv('data/cao2019_20211101_213010.csv', sep='\t')
df_2019

Course Code                             INSTITUTION and COURSE   EOS   \
0         AL801    Software Design with Virtual Reality and Gaming   304    
1         AL802               Software Design with Cloud Computing   301    
2         AL803   Software Design with Mobile Apps and Connected...  309    
3         AL805        Network Management and Cloud Infrastructure   329    
4         AL810                                 Quantity Surveying   307    
..           ...                                                ...   ...   
925       WD200                                     Arts (options)   221    
926       WD210                       Software Systems Development   271    
927       WD211                                 Creative Computing   275    
928       WD212                    Recreation and Sport Management   274    
929       WD230           Mechanical and Manufacturing Engineering   273    

     Mid   
0    328   
1    306   
2    337   
3    442   
4    349   
..    ...  
925  296   
926  329   
927  322   
928  311   
929  348   

[930 rows x 4 columns]

In [625]:
# Check columns to find white space. 
df_2019.columns

Index(['Course Code ', 'INSTITUTION and COURSE ', 'EOS ', 'Mid '], dtype='object')

In [626]:
# Remove white space in column titles.
df_2019.columns = df_2019.columns.str.strip()

As seen above there is white space at the end of the course code and course title strings. It is important to move any white space now as it will impede access to the data later on. 


In [627]:
# Strip data of any white space.
df_2019['Course Code'] = df_2019['Course Code'].str.strip()
df_2019['INSTITUTION and COURSE'] = df_2019['INSTITUTION and COURSE'].str.strip()

## Concatenate & Join Data Sets
***

In [628]:
# Reset index.
df_2021 = df_2021.reset_index()

In [629]:
courses2021 = df_2021[['course_code', 'course_title']]

In [630]:
courses2021

course_code                                       course_title
0         AL801  Software Design for Virtual Reality and Gaming...
1         AL802  Software Design in Artificial Intelligence for...
2         AL803  Software Design for Mobile Apps and Connected ...
3         AL805  Computer Engineering for Network Infrastructur...
4         AL810  Quantity Surveying                            ...
..          ...                                                ...
944       WD211  Creative Computing                            ...
945       WD212  Recreation and Sport Management               ...
946       WD230  Mechanical and Manufacturing Engineering      ...
947       WD231  Early Childhood Care and Education            ...
948       WD232  Business Information Systems                  ...

[949 rows x 2 columns]

In [631]:
# Create new data frame with columns that are applicable to our analysis.
courses2020 = df_2020[['COURSE CODE2', 'COURSE TITLE']]
courses2020.columns = ['course_code', 'course_title']

In [632]:
courses2020

course_code                                       course_title
0          AC120                             International Business
1          AC137                                       Liberal Arts
2          AD101   First Year Art & Design (Common Entry,portfolio)
3          AD102  Graphic Design and Moving Image Design (portfo...
4          AD103  Textile & Surface Design and Jewellery & Objec...
...          ...                                                ...
1459       WD208                          Manufacturing Engineering
1460       WD210                       Software Systems Development
1461       WD211                                 Creative Computing
1462       WD212                    Recreation and Sport Management
1463       WD230           Mechanical and Manufacturing Engineering

[1464 rows x 2 columns]

In [633]:
df_2019.columns

Index(['Course Code', 'INSTITUTION and COURSE', 'EOS', 'Mid'], dtype='object')

In [634]:
courses2019 = df_2019[['Course Code', 'INSTITUTION and COURSE', ]]
courses2019.columns = ['course_code', 'course_title', ]
courses2019

course_code                                       course_title
0         AL801    Software Design with Virtual Reality and Gaming
1         AL802               Software Design with Cloud Computing
2         AL803  Software Design with Mobile Apps and Connected...
3         AL805        Network Management and Cloud Infrastructure
4         AL810                                 Quantity Surveying
..          ...                                                ...
925       WD200                                     Arts (options)
926       WD210                       Software Systems Development
927       WD211                                 Creative Computing
928       WD212                    Recreation and Sport Management
929       WD230           Mechanical and Manufacturing Engineering

[930 rows x 2 columns]

#### Concatenate Data Frames
***

Pandas function to concatenate data frames: 

[Documentation](https://pandas.pydata.org/docs/reference/api/pandas.concat.html) 

&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
`pandas.concat(objs, axis=0, join='outer', ignore_index=False, keys=None, levels=None,` 

&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;`names=None, verify_integrity=False, sort=False, copy=True)`&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;

Setting `ignore_index` to True ensures that the indices from the old data frames are not brought into the new data frame. This is to ensure no duplication of indices. For example, we do not want multiple rows all with 0 as their index.

In [635]:
# Concatenating 2021 & 2021 courses. 
all_courses = pd.concat([courses2021, courses2020, courses2019], ignore_index=True)
all_courses

course_code                                       course_title
0          AL801  Software Design for Virtual Reality and Gaming...
1          AL802  Software Design in Artificial Intelligence for...
2          AL803  Software Design for Mobile Apps and Connected ...
3          AL805  Computer Engineering for Network Infrastructur...
4          AL810  Quantity Surveying                            ...
...          ...                                                ...
3338       WD200                                     Arts (options)
3339       WD210                       Software Systems Development
3340       WD211                                 Creative Computing
3341       WD212                    Recreation and Sport Management
3342       WD230           Mechanical and Manufacturing Engineering

[3343 rows x 2 columns]

In [636]:
# Reset index.
all_courses.reset_index()

index course_code                                       course_title
0         0       AL801  Software Design for Virtual Reality and Gaming...
1         1       AL802  Software Design in Artificial Intelligence for...
2         2       AL803  Software Design for Mobile Apps and Connected ...
3         3       AL805  Computer Engineering for Network Infrastructur...
4         4       AL810  Quantity Surveying                            ...
...     ...         ...                                                ...
3338   3338       WD200                                     Arts (options)
3339   3339       WD210                       Software Systems Development
3340   3340       WD211                                 Creative Computing
3341   3341       WD212                    Recreation and Sport Management
3342   3342       WD230           Mechanical and Manufacturing Engineering

[3343 rows x 3 columns]

## Manage Duplicate Rows
***

To deal with duplicate rows, the following functions form the pandas library can be used: 

&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;`pandas.DataFrame.duplicated()` [Documentation](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.duplicated.html)

&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;`pandas.DataFrame.drop_duplicates()` [Documentation](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.drop_duplicates.html)

In [637]:
# Find extra duplicates based on course code. 
all_courses[all_courses.duplicated(subset=['course_code'])]

course_code                                       course_title
949        AC120                             International Business
950        AC137                                       Liberal Arts
951        AD101   First Year Art & Design (Common Entry,portfolio)
952        AD102  Graphic Design and Moving Image Design (portfo...
953        AD103  Textile & Surface Design and Jewellery & Objec...
...          ...                                                ...
3338       WD200                                     Arts (options)
3339       WD210                       Software Systems Development
3340       WD211                                 Creative Computing
3341       WD212                    Recreation and Sport Management
3342       WD230           Mechanical and Manufacturing Engineering

[1692 rows x 2 columns]

In [638]:
# Remove duplicate rows, implementing changes in-place, while ignoring indices. 
all_courses.drop_duplicates(subset=['course_code'], inplace=True, ignore_index=True)

In [639]:
# Take a look at the data. 
all_courses

course_code                                       course_title
0          AL801  Software Design for Virtual Reality and Gaming...
1          AL802  Software Design in Artificial Intelligence for...
2          AL803  Software Design for Mobile Apps and Connected ...
3          AL805  Computer Engineering for Network Infrastructur...
4          AL810  Quantity Surveying                            ...
...          ...                                                ...
1646       SG441                              Environmental Science
1647       SG446                                Applied Archaeology
1648       TL803                                   Music Technology
1649       TL812                       Computing with Digital Media
1650       WD148                                  Tourism Marketing

[1651 rows x 2 columns]

In [640]:
# Compare two rows.
all_courses.loc[175] == all_courses.loc[176]

course_code     False
course_title    False
dtype: bool

<br>

### Join the points
***

This code joins the 2021 points to the all_course dataframe. 

In [641]:
df_2021

course_code                                       course_title rnd_1  \
0         AL801  Software Design for Virtual Reality and Gaming...   300   
1         AL802  Software Design in Artificial Intelligence for...   313   
2         AL803  Software Design for Mobile Apps and Connected ...   350   
3         AL805  Computer Engineering for Network Infrastructur...   321   
4         AL810  Quantity Surveying                            ...   328   
..          ...                                                ...   ...   
944       WD211  Creative Computing                            ...   270   
945       WD212  Recreation and Sport Management               ...   262   
946       WD230  Mechanical and Manufacturing Engineering      ...   230   
947       WD231  Early Childhood Care and Education            ...   266   
948       WD232  Business Information Systems                  ...   261   

    rnd_2 EOS Midpoints Test/Interview  
0     NaN           359            NaN  
1     NaN           381            NaN  
2     NaN           398            NaN  
3     NaN           381            NaN  
4     NaN           377            NaN  
..    ...           ...            ...  
944   NaN           392            NaN  
945   NaN           304            NaN  
946   230           361            NaN  
947   NaN           366            NaN  
948   261           329            NaN  

[949 rows x 6 columns]

### Why are we setting the index to be column code? 

So that we can join the rows based on common course_code and not on index.

In [642]:
# Change df_2021 column names to include the year.
df_2021.columns = ['course_code', 'course_title', 'points_r1_2021', 'points_r2_2021', 'eos_mid_2021', 'test/interview_2021']

# Check dataframe.
df_2021

course_code                                       course_title  \
0         AL801  Software Design for Virtual Reality and Gaming...   
1         AL802  Software Design in Artificial Intelligence for...   
2         AL803  Software Design for Mobile Apps and Connected ...   
3         AL805  Computer Engineering for Network Infrastructur...   
4         AL810  Quantity Surveying                            ...   
..          ...                                                ...   
944       WD211  Creative Computing                            ...   
945       WD212  Recreation and Sport Management               ...   
946       WD230  Mechanical and Manufacturing Engineering      ...   
947       WD231  Early Childhood Care and Education            ...   
948       WD232  Business Information Systems                  ...   

    points_r1_2021 points_r2_2021 eos_mid_2021 test/interview_2021  
0              300            NaN          359                 NaN  
1              313            NaN          381                 NaN  
2              350            NaN          398                 NaN  
3              321            NaN          381                 NaN  
4              328            NaN          377                 NaN  
..             ...            ...          ...                 ...  
944            270            NaN          392                 NaN  
945            262            NaN          304                 NaN  
946            230            230          361                 NaN  
947            266            NaN          366                 NaN  
948            261            261          329                 NaN  

[949 rows x 6 columns]

In [643]:
# Join the 2021 points to the all_courses dataframe.
all_courses = all_courses.join(df_2021[['points_r1_2021', 'points_r1_2021', 'eos_mid_2021', 'test/interview_2021']])

In [644]:
# Set the all_courses index to be the course_code column.
all_courses = all_courses.set_index('course_code')

In [645]:
all_courses

course_title points_r1_2021  \
course_code                                                                     
AL801        Software Design for Virtual Reality and Gaming...            300   
AL802        Software Design in Artificial Intelligence for...            313   
AL803        Software Design for Mobile Apps and Connected ...            350   
AL805        Computer Engineering for Network Infrastructur...            321   
AL810        Quantity Surveying                            ...            328   
...                                                        ...            ...   
SG441                                    Environmental Science            NaN   
SG446                                      Applied Archaeology            NaN   
TL803                                         Music Technology            NaN   
TL812                             Computing with Digital Media            NaN   
WD148                                        Tourism Marketing            NaN   

            points_r1_2021 eos_mid_2021 test/interview_2021  
course_code                                                  
AL801                  300          359                 NaN  
AL802                  313          381                 NaN  
AL803                  350          398                 NaN  
AL805                  321          381                 NaN  
AL810                  328          377                 NaN  
...                    ...          ...                 ...  
SG441                  NaN          NaN                 NaN  
SG446                  NaN          NaN                 NaN  
TL803                  NaN          NaN                 NaN  
TL812                  NaN          NaN                 NaN  
WD148                  NaN          NaN                 NaN  

[1651 rows x 5 columns]

Join the 2020 points to the all_courses dataframe.

In [646]:
df_2020.columns

Index(['CATEGORY (i.e.ISCED description)', 'COURSE TITLE', 'COURSE CODE2',
       'R1 POINTS', 'R1 Random *', 'R2 POINTS', 'R2 Random*', 'EOS',
       'EOS Random *', 'EOS Mid-point', 'LEVEL', 'HEI', 'Test/Interview #',
       'avp', 'v', 'Column1', 'Column2', 'Column3', 'Column4', 'Column5',
       'Column6', 'Column7', 'Column8'],
      dtype='object')

In [647]:
# Extract columns. 
df_2020 = df_2020[['COURSE CODE2', 'R1 POINTS', 'R2 POINTS', 'EOS Mid-point', 'Test/Interview #']]

# Change column names.
df_2020.columns = ['course_code', 'points_r1_2020', 'points_r2_2020', 'eos_mid_2020', 'test/interview_2020']
df_2020

course_code points_r1_2020 points_r2_2020 eos_mid_2020  \
0          AC120            209            NaN          280   
1          AC137            252            NaN          270   
2          AD101       #+matric            NaN     #+matric   
3          AD102       #+matric            NaN     #+matric   
4          AD103       #+matric            NaN     #+matric   
...          ...            ...            ...          ...   
1459       WD208            188            NaN          339   
1460       WD210            279            NaN          337   
1461       WD211            271            NaN          318   
1462       WD212            270            NaN          349   
1463       WD230            253            NaN          369   

     test/interview_2020  
0                    NaN  
1                    NaN  
2                      #  
3                      #  
4                      #  
...                  ...  
1459                 NaN  
1460                 NaN  
1461                 NaN  
1462                 NaN  
1463                 NaN  

[1464 rows x 5 columns]

In [648]:
# Set index to be course_code column. 
df_2020.set_index('course_code', inplace=True)

In [649]:
# Join 2020 points to all_courses dataframe.
all_courses = all_courses.join(df_2020)

# Check that the points were added.
all_courses

course_title points_r1_2021  \
course_code                                                                     
AL801        Software Design for Virtual Reality and Gaming...            300   
AL802        Software Design in Artificial Intelligence for...            313   
AL803        Software Design for Mobile Apps and Connected ...            350   
AL805        Computer Engineering for Network Infrastructur...            321   
AL810        Quantity Surveying                            ...            328   
...                                                        ...            ...   
SG441                                    Environmental Science            NaN   
SG446                                      Applied Archaeology            NaN   
TL803                                         Music Technology            NaN   
TL812                             Computing with Digital Media            NaN   
WD148                                        Tourism Marketing            NaN   

            points_r1_2021 eos_mid_2021 test/interview_2021 points_r1_2020  \
course_code                                                                  
AL801                  300          359                 NaN            303   
AL802                  313          381                 NaN            332   
AL803                  350          398                 NaN            337   
AL805                  321          381                 NaN            333   
AL810                  328          377                 NaN            319   
...                    ...          ...                 ...            ...   
SG441                  NaN          NaN                 NaN            NaN   
SG446                  NaN          NaN                 NaN            NaN   
TL803                  NaN          NaN                 NaN            NaN   
TL812                  NaN          NaN                 NaN            NaN   
WD148                  NaN          NaN                 NaN            NaN   

            points_r2_2020 eos_mid_2020 test/interview_2020  
course_code                                                  
AL801                  NaN          367                 NaN  
AL802                  NaN          382                 NaN  
AL803                  NaN          360                 NaN  
AL805                  NaN          360                 NaN  
AL810                  NaN          352                 NaN  
...                    ...          ...                 ...  
SG441                  NaN          NaN                 NaN  
SG446                  NaN          NaN                 NaN  
TL803                  NaN          NaN                 NaN  
TL812                  NaN          NaN                 NaN  
WD148                  NaN          NaN                 NaN  

[1651 rows x 9 columns]

In [650]:
df_2019.columns

Index(['Course Code', 'INSTITUTION and COURSE', 'EOS', 'Mid'], dtype='object')

In [651]:
# Extract columns. 
df_2019 = df_2019[['Course Code', 'Mid', 'EOS']]

# Change column names.
df_2019.columns = ['course_code', 'eos_mid_2019', 'eos_2019']
df_2019

course_code eos_mid_2019 eos_2019
0         AL801         328      304 
1         AL802         306      301 
2         AL803         337      309 
3         AL805         442      329 
4         AL810         349      307 
..          ...          ...      ...
925       WD200         296      221 
926       WD210         329      271 
927       WD211         322      275 
928       WD212         311      274 
929       WD230         348      273 

[930 rows x 3 columns]

In [652]:
# Set the index for df_2019 to be the course_code column. 
df_2019.set_index('course_code', inplace=True)

In [653]:
# Join 2019 points to all_courses dataframe.
all_courses = all_courses.join(df_2019)

# Check.
all_courses

course_title points_r1_2021  \
course_code                                                                     
AL801        Software Design for Virtual Reality and Gaming...            300   
AL802        Software Design in Artificial Intelligence for...            313   
AL803        Software Design for Mobile Apps and Connected ...            350   
AL805        Computer Engineering for Network Infrastructur...            321   
AL810        Quantity Surveying                            ...            328   
...                                                        ...            ...   
SG441                                    Environmental Science            NaN   
SG446                                      Applied Archaeology            NaN   
TL803                                         Music Technology            NaN   
TL812                             Computing with Digital Media            NaN   
WD148                                        Tourism Marketing            NaN   

            points_r1_2021 eos_mid_2021 test/interview_2021 points_r1_2020  \
course_code                                                                  
AL801                  300          359                 NaN            303   
AL802                  313          381                 NaN            332   
AL803                  350          398                 NaN            337   
AL805                  321          381                 NaN            333   
AL810                  328          377                 NaN            319   
...                    ...          ...                 ...            ...   
SG441                  NaN          NaN                 NaN            NaN   
SG446                  NaN          NaN                 NaN            NaN   
TL803                  NaN          NaN                 NaN            NaN   
TL812                  NaN          NaN                 NaN            NaN   
WD148                  NaN          NaN                 NaN            NaN   

            points_r2_2020 eos_mid_2020 test/interview_2020 eos_mid_2019  \
course_code                                                                
AL801                  NaN          367                 NaN         328    
AL802                  NaN          382                 NaN         306    
AL803                  NaN          360                 NaN         337    
AL805                  NaN          360                 NaN         442    
AL810                  NaN          352                 NaN         349    
...                    ...          ...                 ...          ...   
SG441                  NaN          NaN                 NaN         358    
SG446                  NaN          NaN                 NaN         290    
TL803                  NaN          NaN                 NaN         288    
TL812                  NaN          NaN                 NaN         369    
WD148                  NaN          NaN                 NaN         338    

            eos_2019  
course_code           
AL801           304   
AL802           301   
AL803           309   
AL805           329   
AL810           307   
...              ...  
SG441           297   
SG446           289   
TL803           264   
TL812           369   
WD148           269   

[1651 rows x 11 columns]

At this stage it would be a good idea to do a spot check to ensure that the joins took place correctly and the data maintained its integrity.

In [654]:
df_2019['mid_2019']

KeyError: 'mid_2019'

***
# End